### ISSUE TITLE
BUG: strange behavior for 1d array In-Place Modification `a[:]`

### ISSUE THREAD

**Issue by wbvguo** [numpy/numpy#25904]
### Describe the issue:

Hi,

I found that in-place modificaiton `[:]` for 1d numpy array `a` is not working (with values automatically set to be 0) in the following code, while it works fine for the 2d numpy array `A`. This is  quite strange to me, as I thought in each iteration, I already created a list `[p[0], p[1]]` and used it to modify `a` in place, however it tends out no success. 

I am not sure if this behavior is intended. But that the in-place modificaiton doesn't work for 1d array but works for 2d array is quite confusing.

![image](https://github.com/numpy/numpy/assets/31968837/4db9b99e-57b2-400b-bf9d-120ccd94067f)

### Reproduce the code example:

```python
A = np.zeros((2,2)) # 2d array with 2x2 dimension
a = np.array([0,1]) # 1d array with length 2

for _ in range(2):
    p = np.random.rand(2) # generate random number
    print("p=", p)
    
    print(f"[ Before ] A's id {id(A)}; A's value {list(A)}")
    A[:] = [[1 - p[0], p[0]], [p[1], 1 - p[1]]]
    print(f"[ After ] A's id {id(A)}; A's value {list(A)}")
    
    
    print(f"[ Before ] a's id {id(a)}; a's value {a}")
    a[:] = [p[0], p[1]]  # doesn't work
    #a[:] = p            # doesn't work
    #a[:] = np.copy(p)   # doesn't work
    #a = p               # work
    #a = np.copy(p)      # work
    #a = [p[0], p[1]]    # work
    print(f"[ After ] a's id {id(a)}; a's value {a}")
```

### Python and NumPy Versions:

1.21.5
3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]

**Comment by jakevdp** [comment]
Hi - it looks like this is behaving as expected. The problem here is that `a` is an integer array, so for in-place updates, updated values will be cast to integer type:
```python
>>> import numpy as np
>>> a = np.array([0, 1])
>>> print(a.dtype)
int64
>>> p = np.random.rand(2)
>>> print(p)
[0.49533001 0.93210754]
>>> a[:] = p  # implicit integer cast
>>> print(a)
[0 0]
>>> print(p.astype(a.dtype))  # these are the values you are updating
[0 0]
```
If you want to avoid this kind of casting when updating values in-place, you should ensure that `a` has a floating point type:
```
>>> a = np.array([0, 1], dtype=float)
>>> a[:] = p
>>> print(a)
[0.49533001 0.93210754]
```
The reason it works for `A` in your example is not because it is two-dimensional, but because `A` has a floating-point type.

Hope that helps!

**Comment by wbvguo** [comment]
@jakevdp, that makes sense; thanks for the help! 

I would also like to provide some background information here in case other users with a similar problem want to get some hints.

**Background**:
I was implementing an iterative optimization algorithm where the parameters to be optimized are floats and always in the (0,1) range. When testing, I found the code ran without a problem, but the parameters never updated. I spent quite some time trying to pinpoint and debug. It turns out the algorithm implementation was fine, and changing the "in-place update" to "assign" solved the no-update issue. That's how I spotted this behavior ( `a = np.array([0,1])`  has default type int, leading to no update via implicit casting).

It looks like, whenever we create an array, explicitly specifying the correct data type is always a good habit. 

So when you use the in-place update and find sth unexpected, perhaps checking the array data type is helpful
